### PR TITLE
Allow more options to the Apollo Server config

### PR DESCRIPTION
**Pull request by abdonrd**
#### Description of what you did:

Allow more options to the Apollo Server config

### CODE REVIEW

File: packages/strapi-plugin-graphql/hooks/graphql/index.js
Comment (alexandrebodin) [inline packages/strapi-plugin-graphql/hooks/graphql/index.js:109]: you can put those as defaults in the serverParams to make it even simpler

**Review by alexandrebodin — CHANGES_REQUESTED** [review]
Yeah sth like that would allow to pass any param the users want. In v4 we can break the config to have a apolloServerConfig key to make it cleaner :)

File: packages/strapi-plugin-graphql/hooks/graphql/index.js
Comment (abdonrd) [reply packages/strapi-plugin-graphql/hooks/graphql/index.js]: Done!

File: packages/strapi-plugin-graphql/hooks/graphql/index.js
Comment (alexandrebodin) [inline packages/strapi-plugin-graphql/hooks/graphql/index.js:127]: you can directly set tracing: false as it will be overwritten byt the serverConfig, some goes for introspection and engine :)

Comment (abdonrd) [reply packages/strapi-plugin-graphql/hooks/graphql/index.js:127]: Oops! You are totally right! Rush!

**Comment by alexandrebodin** [comment]
sth wrong with the tests :)

**Review by alexandrebodin — COMMENTED** [review]
this will need a doc update before we can merge and a deprecation warning :)

**Comment by abdonrd** [comment]
@alexandrebodin all updated! 🚀 

How would you want to add a deprecation warning?

File: packages/strapi-plugin-graphql/hooks/graphql/index.js
Comment (alexandrebodin) [inline packages/strapi-plugin-graphql/hooks/graphql/index.js:99]: can you name this `deprecatedApolloServerConfig`

File: packages/strapi-plugin-graphql/hooks/graphql/index.js
Comment (alexandrebodin) [inline packages/strapi-plugin-graphql/hooks/graphql/index.js:108]: empty object here so we realy separate the two

**Review by alexandrebodin — CHANGES_REQUESTED** [review]
@derrickmehaffy can you help @abdonrd to update the graphql doc with a deprecation warning for those old params

File: packages/strapi-plugin-graphql/hooks/graphql/index.js
Comment (abdonrd) [reply packages/strapi-plugin-graphql/hooks/graphql/index.js]: Done!

File: packages/strapi-plugin-graphql/hooks/graphql/index.js
Comment (abdonrd) [reply packages/strapi-plugin-graphql/hooks/graphql/index.js]: Done!

**Comment by alexandrebodin** [comment]
> @alexandrebodin all updated! 🚀
> 
> How would you want to add a deprecation warning?


I think we should remove the old keys from the default config. and if the user sets them display a warning with strapi.log.warn()

**Comment by derrickmehaffy** [comment]
@alexandrebodin @abdonrd sure, we are just moving the tracing one right?

**Comment by abdonrd** [comment]
> @alexandrebodin @abdonrd sure, we are just moving the tracing one right?

Yeah!

**Comment by derrickmehaffy** [comment]
Would be best to just adjust the documentation and move this into a major release and throw the warning in the migration guide. It's small, and the number of people actually using the tracing feature should be very low as it decreases the performance of GraphQL quite a lot.

If you want to write something up here @abdonrd I'll just ensure it gets added to the next migration guide.

@alexandrebodin did we have a planned date for `v3.2.x` ?

**Comment by abdonrd** [comment]
@derrickmehaffy major release? This is not a breaking change 😕

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy major release? This is not a breaking change

It will be for those that are using the tracing feature no? :wink: Any time we start restructuring the config, that constitutes a breaking change as you can't just change the version numbers within the `package.json` and be running fine, it requires you to adjust the config.

**Comment by abdonrd** [comment]
But the `tracing` still works. It is just marked as deprecated in favor of the new `apolloServerConfig` option.

And it will be removed in the next major version.

**Comment by derrickmehaffy** [comment]
Ah alright I see, we allowing for both (first time I've ever seen Alex allow that :stuck_out_tongue_closed_eyes: ) In that case lets stick a warning here: `docs/v3.x/plugins/graphql.md` Around line 52 just above the config.

something along the lines of 
```
::: warning
Please note the setting for GraphQL tracing as changed and has been moved to `whatever the path is`
:::
```

**Comment by abdonrd** [comment]
@derrickmehaffy  Done!

File: docs/v3.x/plugins/graphql.md
Comment (derrickmehaffy) [inline docs/v3.x/plugins/graphql.md:54]: `apolloServerConfig.tracing`

**Comment by derrickmehaffy** [comment]
Reminder to myself when we start looking at redoing the docs, this entire GraphQL documentation needs to be rewritten x_x

**Review by derrickmehaffy — DISMISSED** [review]
LGTM on the docs side, pending review from Alex on the code side

File: packages/strapi-plugin-graphql/hooks/graphql/index.js
Comment (alexandrebodin) [inline packages/strapi-plugin-graphql/hooks/graphql/index.js:106]: you will have to check with a has instead or the warning won't display if the vars arf falsy

Comment (abdonrd) [reply packages/strapi-plugin-graphql/hooks/graphql/index.js]: Ops! You'r right! Fixed!

File: packages/strapi-plugin-graphql/hooks/graphql/index.js
Comment (alexandrebodin) [inline packages/strapi-plugin-graphql/hooks/graphql/index.js:105]: Nice try but this will check if you have strapi.plugins.graphql.config.tracing.introspection.engine here :) the array is considered as a path :) you will need 3 has ;)

Comment (abdonrd) [reply packages/strapi-plugin-graphql/hooks/graphql/index.js:105]: Ouch! I assumed this would work, but you are totally rigth!

**Comment by abdonrd** [comment]
Thanks for the review @alexandrebodin! It should work now!

**Review by alexandrebodin — APPROVED** [review]

**Pull request merged**
